### PR TITLE
STM32 Bootloader

### PR DESCRIPTION
**Pull request by rickyelopez**
Getting started on a bootloader for the STM32F103

So far, we're at around 7k of flash used for this. The original stm32duino bootloader (which used USB and DFU) was at about 8k.

To-do list:
- [x] get bootloader to boot
- [x] get bootloader to boot an app
- [x] get bootloader to talk on CAN 
- [x] bring up ISOTP
- [x] bring up UDS
- [ ] bring up UDS services
  - [x] tester present
  - [x] ecu reset
  - [x] routine control
  - [ ] download sequence
- [x] write flash manipulation code
- [x] write CRC calculation and comparison code
- [ ] Build a bootloader updater so that the bootloader can also be updated over CAN (this can technically actually happen separately

Not to mention, the other side of this which is to build out a full suite of tooling to actually leverage these capabilities on a computer. I have started writing a bit of code for this, since I needed it to test some of the above functions. There's a LOT to do here though.

### CODE REVIEW

File: components/bootloaders/STM/stm32f103/include/CAN.h
Comment (JoshLafleur) [inline components/bootloaders/STM/stm32f103/include/CAN.h:1]: (nit) Needs file documentation and headers

File: components/bootloaders/STM/stm32f103/include/HW_CAN.h
Comment (JoshLafleur) [inline components/bootloaders/STM/stm32f103/include/HW_CAN.h:223]: Keep consistent capitalization *_T

File: components/bootloaders/STM/stm32f103/include/HW.h
Comment (JoshLafleur) [inline components/bootloaders/STM/stm32f103/include/HW.h:133]: Consistent naming

File: components/bootloaders/STM/stm32f103/include/HW.h
Comment (JoshLafleur) [inline components/bootloaders/STM/stm32f103/include/HW.h:231]: Consistent naming

File: components/bootloaders/STM/stm32f103/include/HW_GPIO.h
Comment (JoshLafleur) [inline components/bootloaders/STM/stm32f103/include/HW_GPIO.h:38]: Consistent naming

File: components/bootloaders/STM/stm32f103/include/HW_GPIO.h
Comment (JoshLafleur) [inline components/bootloaders/STM/stm32f103/include/HW_GPIO.h:44]: Consistent naming

File: components/bootloaders/STM/stm32f103/include/HW_GPIO.h
Comment (JoshLafleur) [inline components/bootloaders/STM/stm32f103/include/HW_GPIO.h:54]: Consistent naming

File: components/bootloaders/STM/stm32f103/include/Utilities.h
Comment (JoshLafleur) [inline components/bootloaders/STM/stm32f103/include/Utilities.h:10]: Reading material:
https://lxr.linux.no/#linux+v2.6.31/include/linux/compiler-gcc.h#L12
https://en.wikipedia.org/wiki/Memory_ordering
Discussion: https://stackoverflow.com/questions/14950614/working-of-asm-volatile-memory

It seems it stops memory access ordering optimization at compile time. Helps make a stricter memory access flow, however does not prevent against CPU instruction optimization

File: components/bootloaders/STM/stm32f103/include/Utilities.h
Comment (JoshLafleur) [inline components/bootloaders/STM/stm32f103/include/Utilities.h:1]: Have the Atomic instructions been tested?

File: components/heartbeat/STM32F103C8.ld
Comment (JoshLafleur) [inline components/heartbeat/STM32F103C8.ld:16]: I think we would benefit from allocation a 1kB nvm area at the back of flash. Will be useful to integrate it from the start for all of the bms and other systems

**Review by JoshLafleur — COMMENTED** [review]
Files needs headers and documentation, but I definetly like the idea of going towards fully custom firmware custom, really great stuff going and getting all of that laid out!

I like the starting point and direction, really awesome of you to take that lead

**Comment by rickyelopez** [comment]
> Files needs headers and documentation, but I definetly like the idea of going towards fully custom firmware custom, really great stuff going and getting all of that laid out!
> 
> I like the starting point and direction, really awesome of you to take that lead

lots of cleanup to be done before this is anywhere near merge-able (not the least of which is documentation), namely getting it to actually do bootloader things lol

File: components/bootloaders/STM/stm32f103/include/CAN.h
Comment (rickyelopez) [reply components/bootloaders/STM/stm32f103/include/CAN.h:1]: yeah all these files are going to need that

File: components/bootloaders/STM/stm32f103/include/HW_CAN.h
Comment (rickyelopez) [reply components/bootloaders/STM/stm32f103/include/HW_CAN.h:223]: I think we should stick to lowercase t, like with all the standard types (i.e `uint8_t`)

File: components/bootloaders/STM/stm32f103/include/HW.h
Comment (rickyelopez) [reply components/bootloaders/STM/stm32f103/include/HW.h:133]: yeah this stuff is all copy-pasted, I'll rename everything before it's ready to merge. More important to get it working right now though

File: components/bootloaders/STM/stm32f103/include/Utilities.h
Comment (rickyelopez) [reply components/bootloaders/STM/stm32f103/include/Utilities.h:10]: nice, very interesting

File: components/bootloaders/STM/stm32f103/include/Utilities.h
Comment (rickyelopez) [reply components/bootloaders/STM/stm32f103/include/Utilities.h:1]: They come straight from the cortex-m3 library files, they're intrinsics. We'll have to test them to confirm, though

File: components/heartbeat/STM32F103C8.ld
Comment (rickyelopez) [reply components/heartbeat/STM32F103C8.ld:16]: once the bootloader size is stabilized we can allocate an extra page at the beginning of the flash for nvm. iirc the page sizes are bigger at the end of flash than at the beginning, so ideally we want to allocate them at the beginning of flash for the bootloader so they don't consume too much application space. The apps can use whatever is left at the end of flash for nvm, unless there's less than a page. If we run into that, we could also explore having the app's nvm pages at the front too

**Comment by rickyelopez** [comment]
Really cool progress!

image

CAN messages being sent out from four different periodic functions which are driven by the timer. Also, UDS frame being sent out (just a simple Tester Present ACK) by the library I started writing, and you can see it is decoded correctly as a UDS frame!

**Comment by rickyelopez** [comment]
This PR is starting to get a little big, and it will only get worse. I propose we merge this as-is since most (if not all) of the behavior is confirmed working, and then keep working on it in a subsequent PR.

**Comment by rickyelopez** [comment]
More good progress

image

Got a full UDS transaction to work!
`Send from PC -> receive in the bootloader -> respond -> PC recognizes response`

Cool!

**Comment by rickyelopez** [comment]
With this latest push, app CRC generation (at build time) and verification (in the bootloader) are working!!

image

In the above photo, you can see the periodics spitting out data, but also on address 0x402 you can see the CRC as read from the app's flash region, as well as the app CRC as calculated by the bootloader. As you can see, they match! Very cool!

I think the only big part of this that is left to do is the actual app download, the verification of which will be hugely facilitated by having a functional CRC comparison!

File: components/heartbeat/STM32F103C8.ld
Comment (rickyelopez) [reply components/heartbeat/STM32F103C8.ld:16]: I was wrong about this. On this chip, all flash pages are 1k. We can allocate wherever, but probably not worth doing until we actually have a need for nvm

**Comment by rickyelopez** [comment]
Ok, this is ready to merge. Almost everything we need, with the notable exception of app downloads, is working. Let's merge this so we can progress with other work. I'll do app downloads in a separate pr

**Review by JoshLafleur — APPROVED** [review]

**Pull request merged**
